src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,196 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        public Node(T value) {
+            this.value = value;
+        }
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    public T getHeadValue() {
+        return head.value;
+    }
+
+    public T getTailValue() {
+        return tail.value;
+    }
+
     @Override
     public void add(T value) {
+        insertNodeToTail(value);
     }
 
     @Override
     public void add(T value, int index) {
+        validateIndex(index, 1);
+        if (index == size) {
+            insertNodeToTail(value);
+            return;
+        }
+        if (index == 0) {
+            insertNodeToHead(value);
+            return;
+        }
+        insertNode(findNode(index), value);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T data : list) {
+            add(data);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index, 0);
+        if (index == 0) {
+            return getHeadValue();
+        }
+        if (index == size - 1) {
+            return getTailValue();
+        }
+        return findNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index, 0);
+        final T val;
+        if (index == 0) {
+            val = head.value;
+            head.value = value;
+            return val;
+        }
+        if (index == size - 1) {
+            val = tail.value;
+            tail.value = value;
+            return val;
+        }
+        Node<T> node = findNode(index);
+        val = node.value;
+        node.value = value;
+        return val;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index, 0);
```

Comment:
This is code repeating. You always check the index before getting node by index. What you should do is move the check logic INSIDE the findNode method.